ext/com_dotnet/tests/comtest/comtest.cpp

```diff
@@ -0,0 +1,318 @@
+
+	// IPersistStream
+	STDMETHODIMP IsDirty( void);
+	STDMETHODIMP Load(IStream *pStm);
+	STDMETHODIMP Save(IStream *pStm, BOOL fClearDirty);
+	STDMETHODIMP GetSizeMax(ULARGE_INTEGER *pcbSize);
+
+	CDocument() : m_content(SysAllocString(L"")), m_cRef(1) { g_cComponents++; }
+	~CDocument() { SysFreeString(m_content); g_cComponents--; }
+	HRESULT Init(void);
+
+private:
+	BSTR m_content;
+	ULONG m_cRef;
+	ITypeInfo* m_pTypeInfo;
+	BOOL m_dirty;
+};
+
+ULONG CDocument::AddRef()
+{
+	return ++m_cRef;
+}
+
+ULONG CDocument::Release()
+{
+	if (--m_cRef != 0) {
+		return m_cRef;
+	}
+	m_pTypeInfo->Release();
+	delete this;
+	return 0;
+}
+
+HRESULT CDocument::QueryInterface(REFIID riid, void** ppv)
+{
+	if (riid == IID_IUnknown) {
+		*ppv = (IDocument*) this;
+	} else if (riid == IID_IDocument) {
+		*ppv = (IDocument *) this;
+	} else if (riid == IID_IDispatch) {
+		*ppv = (IDispatch*) this;
+	} else if (riid == IID_IPersistStream) {
+		*ppv = (IPersistStream *) this;
+	} else {
+		*ppv = NULL;
+		return E_NOINTERFACE;
+	}
+	AddRef();
+	return S_OK;
+}
+
+HRESULT CDocument::GetTypeInfoCount(UINT* pCountTypeInfo)
+{
+	*pCountTypeInfo = 1;
+	return S_OK;
+}
+
+HRESULT CDocument::GetTypeInfo(UINT iTypeInfo, LCID lcid, ITypeInfo** ppITypeInfo)
+{
+	*ppITypeInfo = NULL;
+	if (iTypeInfo != 0) {
+		return DISP_E_BADINDEX;
+	}
+	m_pTypeInfo->AddRef();
+	*ppITypeInfo = m_pTypeInfo;
+	return S_OK;
+}
+
+HRESULT CDocument::GetIDsOfNames(REFIID riid, LPOLESTR* rgszNames, UINT cNames, LCID lcid, DISPID* rgDispId)
+{
+	if (riid != IID_NULL) {
+		return DISP_E_UNKNOWNINTERFACE;
+	}
+	return DispGetIDsOfNames(m_pTypeInfo, rgszNames, cNames, rgDispId);
+}
+
+HRESULT CDocument::Invoke(DISPID dispIdMember, REFIID riid, LCID lcid, WORD wFlags, DISPPARAMS* pDispParams,
+	VARIANT* pVarResult, EXCEPINFO* pExcepInfo, UINT* puArgErr)
+{
+	if (riid != IID_NULL) {
+		return DISP_E_UNKNOWNINTERFACE;
+	}
+	return DispInvoke(this, m_pTypeInfo, dispIdMember, wFlags, pDispParams, pVarResult, pExcepInfo, puArgErr); 
+}
+
+HRESULT CDocument::get_Content(BSTR* retvalue)
+{
+	*retvalue = SysAllocString(m_content);
+	return S_OK;
+}
+
+HRESULT CDocument::put_Content(BSTR newvalue)
+{
+	SysFreeString(m_content);
+	m_content = SysAllocString(newvalue);
+	return S_OK;
+}
+
+HRESULT CDocument::Init(void)
+{
+	ITypeLib* pTypeLib;
+	if (FAILED(LoadRegTypeLib(LIBID_ComTest, 1, 0, LANG_NEUTRAL, &pTypeLib))) {
+		return E_FAIL;
+	}
+	HRESULT hr = pTypeLib->GetTypeInfoOfGuid(IID_IDocument, &m_pTypeInfo);
+	if (FAILED(hr)) {
+		pTypeLib->Release();
+		return hr;
+	}
+
+	pTypeLib->Release();
+	return S_OK;
+}
+
+HRESULT CDocument::GetClassID(CLSID *pClassID)
+{
+	*pClassID = CLSID_Document;
+	return S_OK;
+}
+
+HRESULT CDocument::IsDirty(void)
+{
+	return m_dirty ? S_OK : S_FALSE;
+}
+
+HRESULT CDocument::Load(IStream *pStm)
+{
+	ULONG read = 0;
+	ULONG cbSize;
+	char *string;
+	if (FAILED(pStm->Read(&cbSize, sizeof cbSize, &read))) {
+		return S_FALSE;
+	}
+	string = (char *) malloc(cbSize);

```

Comment:
seems to me there are case of memleaks with it.

Reply:
Indeed. That should be fixed now. However, it is hard to employ some runtime tooling here (LLVM ASan fails to start the inproc server for some reason; MSVC ASan doesn't support LSan; Windows debug heap requires a couple of modifications to comtest.cpp and Makefile.frag.w32).